Mathlib/Topology/Constructible.lean

```diff
@@ -0,0 +1,345 @@
+    (ht : ∀ i ∈ s, IsRetroCompact (t i)) : IsRetroCompact (s.inf' hs t) := by
+  rw [Finset.inf'_eq_inf]; exact .finsetInf ht
+
+lemma IsRetroCompact.inter_isOpen (hs : IsRetroCompact s) (ht : IsRetroCompact t)
+    (htopen : IsOpen t) : IsRetroCompact (s ∩ t : Set X) :=
+  fun _U hUcomp hUopen ↦ inter_assoc .. ▸ hs (ht hUcomp hUopen) (htopen.inter hUopen)
+
+lemma IsRetroCompact.isOpen_inter (hs : IsRetroCompact s) (ht : IsRetroCompact t)
+    (hsopen : IsOpen s) : IsRetroCompact (s ∩ t : Set X) :=
+  inter_comm .. ▸ ht.inter_isOpen hs hsopen
+
+lemma isRetroCompact_iff_isSpectralMap_subtypeVal :
+    IsRetroCompact s ↔ IsSpectralMap (Subtype.val : s → X) := by
+  refine ⟨fun hs ↦ ⟨continuous_subtype_val, fun t htopen htcomp ↦ ?_⟩, fun hs t htcomp htopen ↦ ?_⟩
+  · rw [IsEmbedding.subtypeVal.isCompact_iff, image_preimage_eq_inter_range,
+      Subtype.range_coe_subtype, setOf_mem_eq, inter_comm]
+    exact hs htcomp htopen
+  · simpa using (hs.isCompact_preimage_of_isOpen htopen htcomp).image continuous_subtype_val
+
+@[stacks 005B]
+lemma IsRetroCompact.image_of_isEmbedding (hs : IsRetroCompact s) (hfemb : IsEmbedding f)
+    (hfcomp : IsRetroCompact (range f)) : IsRetroCompact (f '' s) := by
+  rintro U hUcomp hUopen
+  rw [← image_inter_preimage, ← hfemb.isCompact_iff]
+  refine hs ?_ <| hUopen.preimage hfemb.continuous
+  rw [hfemb.isCompact_iff, image_preimage_eq_inter_range, inter_comm]
+  exact hfcomp hUcomp hUopen
+
+@[stacks 005J] -- Extracted from the proof of `005J`
+lemma IsRetroCompact.preimage_of_isOpenEmbedding {s : Set Y} (hf : IsOpenEmbedding f)
+    (hs : IsRetroCompact s) : IsRetroCompact (f ⁻¹' s) := by
+  rintro U hUcomp hUopen
+  rw [hf.isCompact_iff, image_preimage_inter]
+  exact hs (hUcomp.image hf.continuous) <| hf.isOpenMap _ hUopen
+
+@[stacks 09YE] -- Extracted from the proof of `09YE`
+lemma IsRetroCompact.preimage_of_isClosedEmbedding {s : Set Y} (hf : IsClosedEmbedding f)
+    (hf' : IsCompact (range f)ᶜ) (hs : IsRetroCompact s)  : IsRetroCompact (f ⁻¹' s) := by
+  rintro U hUcomp hUopen
+  have hfUopen : IsOpen (f '' U ∪ (range f)ᶜ) := by
+    simpa [← range_diff_image hf.injective, sdiff_eq, compl_inter, union_comm]
+      using (hf.isClosedMap _ hUopen.isClosed_compl).isOpen_compl
+  have hfUcomp : IsCompact (f '' U ∪ (range f)ᶜ) := (hUcomp.image hf.continuous).union hf'
+  simpa [inter_union_distrib_left, inter_left_comm, inter_eq_right.2 (image_subset_range ..),
+    hf.isCompact_iff, image_preimage_inter] using (hs hfUcomp hfUopen).inter_left hf.isClosed_range
+
+/-! ### Constructible sets -/
+
+/-- A constructible set is a set that can be written as the
+finite union/finite intersection/complement of open retrocompact sets.
+
+In other words, constructible sets form the boolean subalgebra generated by open retrocompact sets.
+-/
+def IsConstructible (s : Set X) : Prop :=
+  s ∈ BooleanSubalgebra.closure {U | IsOpen U ∧ IsRetroCompact U}
```

Comment:
Do we have the fact that the Boolean algebra generated by a set `s` closed under `inf` consists of exactly finite union of sets of the form `a \ b` with a, b in s?
If we take `s` to be the set of open retrocompact sets, then we get the more familiar definition seen on Stacks and Wikipedia.

Reply:
No, see the recent Zulip topic about disjunction normal form

Reply:
This is 9.1.3 in Grothendieck, reference above, and should not be hard to formalize directly if needed.

Reply:
Indeed it's not hard; see #20251. It turns out I need that `s` contains top and bot as well.